nnstreamer/tensor_trainer/tensor_trainer_nntrainer.cc

```diff
@@ -0,0 +1,608 @@
+    nntrainer->train_data->push_count = nntrainer->valid_data->push_count = 0;
+  }
+
+  if (nntrainer->train_data->push_count < nntrainer->num_training_samples) {
+    data = nntrainer->train_data.get();
+    ml_logd("#### T-Data ####");
+  } else if (nntrainer->valid_data->push_count <
+             nntrainer->num_validation_samples) {
+    data = nntrainer->valid_data.get();
+    ml_logd("#### V-Data ####");
+  }
+
+  ml_logd("number of inputs(%" PRId64 ") and labels(%" PRId64 ")",
+          nntrainer->num_inputs, nntrainer->num_labels);
+
+  unsigned int idx = 0, i = 0;
+  i = data->queue_rear;
+  ml_logd("Insert, queue_rear : %d", i);
+  for (auto inputs : data->tensor_data[i].inputs) {
+    ml_logd("inputs : %p", inputs);
+    ml_logd("input[%d]:%p, size:%zd\n", idx, input[idx].data, input[idx].size);
+    std::memcpy(inputs, input[idx].data, input[idx].size);
+    idx++;
+  }
+
+  for (auto labels : data->tensor_data[i].labels) {
+    ml_logd("labels : %p", labels);
+    ml_logd("input[%d]:%p, size:%zd\n", idx, input[idx].data, input[idx].size);
+    std::memcpy(labels, input[idx].data, input[idx].size);
+    idx++;
+  }
+
+  data->push_count++;
+  data->queue_count++;
+  data->queue_rear = (data->queue_rear + 1) % data->queue_size;
+  ml_logd("front:%d, rear:%d, filled:%d", data->queue_front, data->queue_rear,
+          data->queue_count);
+
+  if (data->is_data_wait_locked && data->queue_count > 0) {
+    data->data_wait.notify_one();
+    ml_logd("send signal");
+  }
+
+  if (data->queue_count == data->queue_size) {
+    data->is_data_full_locked = TRUE;
+    ml_logd("locked, data is full");
+    std::unique_lock<std::mutex> lock(data->data_full_lock);
+    data->data_full.wait(lock);
+    ml_logd("unlocked, queue is empty");
+    data->is_data_full_locked = FALSE;
+  }
+
+  ml_logd("(pop/push: %" PRId64 "/%" PRId64 ")", data->pop_count,
+          data->push_count);
+  ml_logd("T-pushed: %" PRId64 "/%" PRId64 ", V-pushed:%" PRId64 "/%" PRId64
+          "\n",
+          nntrainer->train_data->push_count, nntrainer->num_training_samples,
+          nntrainer->valid_data->push_count, nntrainer->num_validation_samples);
+
+  ml_logd("<leaved>");
+  return 0;
+}
+
+int getSample(float **input, float **label, bool *last, void *user_data) {
+
+  auto data = reinterpret_cast<NNTrainer::InputTensorsInfo *>(user_data);
+
+  ml_logd("<called>");
+  ml_logd("(pop/push: %" PRId64 "/%" PRId64 ")", data->pop_count,
+          data->push_count);
+  pid_t pid = getpid();
+  pid_t tid = syscall(SYS_gettid);
+
+  ml_logd("<called>");
+  ml_logd("pid[%d], tid[%d]", pid, tid);
+  ml_logd("front:%d, rear:%d", data->queue_front, data->queue_rear);
+  ml_logd("num_inputs: %" PRId64 ", num_labels: %" PRId64 "", data->num_inputs,
+          data->num_labels);
+
+  unsigned int i = 0;
+  unsigned int idx = data->queue_front;
+  ml_logd("Delete, queue_front: %d", idx);
+
+  for (i = 0; i < data->num_inputs; i++) {
```

Comment:
if this is callback function nntrainer call, then it will  always as one at a time. It means that data->num_inputs are always 1.